src/gnutls_session.cpp

```diff
@@ -11,40 +11,81 @@ namespace oxen::quic
             gnutls_session_get_data2 to be called in hook function after handshake completion
     */
 
-    extern "C"
-    {
-        int anti_replay_db_add_func(void* dbf, time_t exp_time, const gnutls_datum_t* key, const gnutls_datum_t* data)
-        {
-            auto* ep = static_cast<Endpoint*>(dbf);
-            assert(ep);
+    static constexpr auto SESSION_TICKET_HEADER = "GNUTLS SESSION PARAMETERS";
 
-            log::warning(log_cat, "0RTT session resumption is not available; callback is no-op");
-            return 0;
+    int gtls_session_callbacks::server_anti_replay_cb(
+            void* dbf, time_t exp_time, const gnutls_datum_t* key, const gnutls_datum_t* data)
+    {
+        log::debug(log_cat, "{} called", __PRETTY_FUNCTION__);
+        auto* ep = static_cast<Endpoint*>(dbf);
+        assert(ep);
 
-            (void)exp_time;
-            (void)key;
-            (void)data;
-            (void)ep;
+        if (not ep->zero_rtt_enabled())
+            throw std::runtime_error{"Anti-replay DB hook should not be called on 0rtt-disabled endpoint!"};
 
-            // return ep->validate_anti_replay({key->data, key->size}, {data->data, data->size}, exp_time);
-        }
+        return ep->validate_anti_replay({key, data}, exp_time);
+    }
 
-        int client_hook_func(
-                gnutls_session_t session,
-                unsigned int htype,
-                unsigned /* when */,
-                unsigned int /* incoming */,
-                const gnutls_datum_t* /* msg */)
+    int gtls_session_callbacks::client_session_cb(
+            gnutls_session_t session,
+            unsigned int htype,
+            unsigned /* when */,
+            unsigned int /* incoming */,
+            const gnutls_datum_t* /* msg */)
+    {
+        if (htype == GNUTLS_HANDSHAKE_NEW_SESSION_TICKET)
         {
-            if (htype == GNUTLS_HANDSHAKE_NEW_SESSION_TICKET)
+            log::debug(log_cat, "Client received new session ticket from server!");
+            auto* conn = get_connection_from_gnutls(session);
+            auto remote_key = conn->remote_key();
+            auto& ep = conn->endpoint();
+            gnutls_datum_t data{}, encoded{};
+
+            if (auto rv = gnutls_session_get_data2(session, &data); rv != 0)
```

Comment:
This is leaking memory: this gnutls function allocates memory that we are responsible for freeing via `gnutls_free`.